File: backend/app/models/knowledge.py
import enum
import uuid
from datetime import datetime

from pgvector.sqlalchemy import Vector
from sqlalchemy import DateTime, Enum, ForeignKey, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base
from app.models.document import EMBEDDING_DIM


class KnowledgeKind(str, enum.Enum):
    """Modelled on Chat2DB's KnowledgeManagementPromptType. Three kinds because they
    answer three genuinely different questions the schema alone can't:

    term            — what a business word means in this database ("حساب کل" -> _Accs,
                      acc_Level=1). The fix for opaque third-party ERP table names.
    business_logic  — rules that aren't visible in the schema ("مبلغ خالص = tr_Deb -
                      tr_Cre", "اسناد با art_State=3 نهایی‌اند").
    sql_template    — a known-correct query for a recurring question, so the model
                      adapts a verified example instead of inventing joins.
    """

    term = "term"
    business_logic = "business_logic"
    sql_template = "sql_template"


class KnowledgeEntry(Base):
    """A workspace's curated domain knowledge, retrieved by embedding similarity and
    injected into the chat prompt. This is the layer that turns a generic NL-to-SQL
    model into one that understands *this* organization's database: the schema says a
    column is called acc_Level, only a knowledge entry says level 1 means «کل»."""

    __tablename__ = "knowledge_entries"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    workspace_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("workspaces.id", ondelete="CASCADE"), nullable=False, index=True
    )
    # Optional: scope an entry to one connection when the same term means different
    # things in different databases. NULL = applies to the whole workspace.
    db_connection_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("db_connections.id", ondelete="CASCADE"), nullable=True, index=True
    )
    kind: Mapped[KnowledgeKind] = mapped_column(
        Enum(KnowledgeKind, name="knowledge_kind"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(300), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    # Nullable so an entry is still saved (and usable via the always-included term list)
    # when the embedding service is down — degrading to no semantic retrieval beats
    # refusing the write.
    embedding: Mapped[list[float] | None] = mapped_column(Vector(EMBEDDING_DIM), nullable=True)
    created_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

File: frontend/src/api/knowledge.ts
import { api } from './client'

/** Mirrors the backend KnowledgeKind — the three kinds of curated domain knowledge. */
export type KnowledgeKind = 'term' | 'business_logic' | 'sql_template'

export interface KnowledgeEntryDto {
  id: string
  workspace_id: string
  db_connection_id: string | null
  kind: KnowledgeKind
  name: string
  content: string
  created_at: string
  updated_at: string
}

export interface KnowledgeEntryCreate {
  kind: KnowledgeKind
  name: string
  content: string
  db_connection_id?: string | null
}

export const knowledgeApi = {
  list: (workspaceId: string, kind?: KnowledgeKind) =>
    api.get<KnowledgeEntryDto[]>(
      `/workspaces/${workspaceId}/knowledge${kind ? `?kind=${kind}` : ''}`,
    ),
  create: (workspaceId: string, payload: KnowledgeEntryCreate) =>
    api.post<KnowledgeEntryDto>(`/workspaces/${workspaceId}/knowledge`, payload),
  update: (workspaceId: string, id: string, payload: Partial<KnowledgeEntryCreate>) =>
    api.patch<KnowledgeEntryDto>(`/workspaces/${workspaceId}/knowledge/${id}`, payload),
  remove: (workspaceId: string, id: string) =>
    api.delete<void>(`/workspaces/${workspaceId}/knowledge/${id}`),
}

export const KNOWLEDGE_KINDS: {
  value: KnowledgeKind
  label: string
  description: string
  namePlaceholder: string
  contentPlaceholder: string
}[] = [
  {
    value: 'term',
    label: 'اصطلاحات',
    description:
      'معنی واژه‌های سازمانی و اینکه هرکدام به کدام جدول یا ستون نگاشت می‌شوند. مهم‌ترین بخش برای دیتابیس‌هایی که نام جدول‌هایشان گویا نیست.',
    namePlaceholder: 'مثلاً: حساب کل',
    contentPlaceholder:
      'مثلاً: در جدول _Accs، رکوردهایی که acc_Level برابر ۱ دارند. کد حساب در acc_Code و نام آن در acc_Name است.',
  },
  {
    value: 'business_logic',
    label: 'قواعد کسب‌وکار',
    description: 'محاسبات و قواعدی که از روی ساختار جدول‌ها قابل حدس زدن نیستند.',
    namePlaceholder: 'مثلاً: محاسبه مانده حساب',
    contentPlaceholder:
      'مثلاً: مانده هر حساب برابر است با مجموع tr_Deb منهای مجموع tr_Cre در جدول _Transes.',
  },
  {
    value: 'sql_template',
    label: 'نمونه کوئری‌ها',
    description:
      'کوئری‌های درست و تأییدشده برای پرسش‌های پرتکرار. مدل به‌جای ساختن JOIN از صفر، نمونه‌ی درست را تطبیق می‌دهد.',
    namePlaceholder: 'مثلاً: لیست حساب‌های کل با مانده',
    contentPlaceholder: 'SELECT a.acc_Code, a.acc_Name, SUM(t.tr_Deb - t.tr_Cre) AS balance\nFROM ...',
  },
]
